services/dkg/pedersen/mod.go

```diff
@@ -47,88 +47,142 @@ const (
+	if err != nil {
+		return nil, xerrors.Errorf("failed to decode electionID: %v", err)
+	}
+
+	// check that electionID corresponds to a real election
+	proof, err := s.service.GetProof(electionIDBuf)
+	if err != nil {
+		return nil, xerrors.Errorf(
+			"failed to read on the blockchain: %v",
+			err,
+		)
+	}
+	if string(proof.GetValue()) == "" {
+		return nil, xerrors.Errorf(
+			"election %s does not exist",
+			electionID,
+		)
+	}
+
+	// link the actor to an RPC by the election ID
+	h := NewHandler(s.mino.GetAddress(), s.service, handlerData)
+	no := s.mino.WithSegment(electionID)
+	rpc := mino.MustCreateRPC(no, "dkgevoting", h, s.factory)
 
 	a := &Actor{
-		rpc:       mino.MustCreateRPC(s.mino, "dkgevoting", h, s.factory),
-		factory:   s.factory,
-		startRes:  h.startRes,
-		service:   s.service,
-		rosterFac: s.rosterFac,
-		context:   jsonserde.NewContext(),
+		rpc:        rpc,
+		factory:    s.factory,
+		service:    s.service,
+		rosterFac:  s.rosterFac,
+		context:    jsonserde.NewContext(),
+		handler:    h,
+		electionID: electionID,
 	}
 
-	s.actor = a
+	s.actors[electionID] = a
 
 	return a, nil
 }
 
-// GetLastActor ...
-func (s *Pedersen) GetLastActor() (dkg.Actor, error) {
-	if s.actor != nil {
-		return s.actor, nil
-	} else {
-		return nil, xerrors.Errorf("listen has not been called")
+func (s *Pedersen) GetActor(electionIDBuf []byte) (dkg.Actor, error) {
+
+	electionID := hex.EncodeToString(electionIDBuf)
+
+	actor, exists := s.actors[electionID]
+	if exists {
+		return actor, nil
 	}
+
+	return nil, xerrors.Errorf("Listen was not called for electionID %s", electionID)
 }
 
 // Actor allows one to perform DKG operations like encrypt/decrypt a message
 //
 // - implements dkg.Actor
 type Actor struct {
-	rpc       mino.RPC
-	factory   serde.Factory
-	startRes  *state
-	service   ordering.Service
-	rosterFac authority.Factory
-	context   serde.Context
+	rpc        mino.RPC
+	factory    serde.Factory
+	service    ordering.Service
+	rosterFac  authority.Factory
+	context    serde.Context
+	handler    *Handler
+	electionID string
 }
 
-// Setup implement dkg.Actor. It initializes the DKG.
-func (a *Actor) Setup(electionID []byte) (kyber.Point, error) {
-	if a.startRes.Done() {
+// Setup implements dkg.Actor. It initializes the DKG.
+func (a *Actor) Setup() (kyber.Point, error) {
+
+	if a.handler.startRes.Done() {
 		return nil, xerrors.Errorf("startRes is already done, only one setup call is allowed")
 	}
+	electionIDBuf, err := hex.DecodeString(a.electionID)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to decode electionID: %v", err)
+	}
 
-	proof, err := a.service.GetProof(electionID)
+	// check that electionID corresponds to a real election
+	proof, err := a.service.GetProof(electionIDBuf)
 	if err != nil {
-		return nil, xerrors.Errorf("failed to read on the blockchain: %v", err)
+		return nil, xerrors.Errorf(
+			"failed to read on the blockchain: %v",
+			err,
+		)
+	}
+	if proof == nil {
```

Comment:
why would `proof` be nil ? Can `GetProof` not return an error and a nil result ?